File: go/pkg/models/point_of_interest_discovery.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type PointOfInterestDiscovery struct {
	ID                uuid.UUID       `json:"id" gorm:"type:uuid;default:uuid_generate_v4()"`
	CreatedAt         time.Time       `json:"createdAt"`
	UpdatedAt         time.Time       `json:"updatedAt"`
	TeamID            *uuid.UUID      `json:"teamId"`
	Team              *Team           `json:"team"`
	UserID            *uuid.UUID      `json:"userId"`
	User              *User           `json:"user"`
	PointOfInterestID uuid.UUID       `json:"pointOfInterestId"`
	PointOfInterest   PointOfInterest `json:"pointOfInterest"`
}

func (PointOfInterestDiscovery) TableName() string {
	return "point_of_interest_discoveries"
}

File: js/packages/types/src/pointOfInterestDiscovery.ts
export interface PointOfInterestDiscovery {
    id: string;
    createdAt: Date;
    updatedAt: Date;
    teamId: string | undefined;
    userId: string | undefined;
    pointOfInterestId: string;
}

export const hasDiscoveredPointOfInterest = (pointOfInterestId: string, entityId: string, pointOfInterestDiscoveries: PointOfInterestDiscovery[]) => {
    return pointOfInterestDiscoveries
        .filter((pointOfInterestDiscovery) => pointOfInterestDiscovery.pointOfInterestId === pointOfInterestId)
        .some((pointOfInterestDiscovery) => pointOfInterestDiscovery.userId === entityId || pointOfInterestDiscovery.teamId === entityId);
  };